pkg/syncer/upsync/upsync_controller.go

```diff
@@ -0,0 +1,253 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upsync
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	kcpdynamic "github.com/kcp-dev/client-go/dynamic"
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	ddsif "github.com/kcp-dev/kcp/pkg/informer"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+)
+
+const controllerName = "kcp-resource-upsyncer"
+
+// NewUpSyncer returns a new controller which upsyncs, through the Usyncer virtual workspace, downstream resources
+// which are part of the upsyncable resource types (fixed limited list for now), and provide
+// the following labels:
+//   - internal.workload.kcp.io/cluster: <sync target key>
+//   - state.workload.kcp.io/<sync target key>: Upsync
+//
+// and optionally, for cluster-wide resources, the `kcp.io/namespace-locator` annotation
+// filled with the information necessary identify the upstream workspace to upsync to.
+func NewUpSyncer(syncerLogger logr.Logger, syncTargetWorkspace logicalcluster.Name,
```

Comment:
elsewhere we call this ClusterName, not Workspace.

Reply:
Fixed